File: ios/Sources/OtaLiveUpdaterPlugin/OtaLiveUpdaterPlugin.swift
import Foundation
import Capacitor
import ZIPFoundation
import BackgroundTasks
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(OtaLiveUpdaterPlugin)
public class OtaLiveUpdaterPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OtaLiveUpdaterPlugin"
    public let jsName = "OtaLiveUpdater"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkpoint",returnType: CAPPluginReturnPromise)
    ]
    private var currentVersion = "1.0.0"
    private var pendingVersionPath: String?
    private let storageDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
    private let taskIdentifier = "team.devapp.capacitor.otalive.versionCheck"
    private var calledCheckpoints: Set<String> = []

    struct OTAUpdate: Codable {
        let version: String
        let date: String
        let releaseDate: String
        let bundleUrl: String
        let description: String
        let sha256: String
    }

    struct UpdateConfig: Codable {
        let requiredCheckpoints: [String: Int]
        let timeoutCheckMillis: Int
    }

    public override func load() {
        registerBackgroundTask()
        scheduleVersionCheck()
    }

    private func registerBackgroundTask() {
        BGTaskScheduler.shared.register(forTaskWithIdentifier: taskIdentifier, using: nil) { task in
            self.handleVersionCheck(task: task as! BGAppRefreshTask)
        }
    }

    private func scheduleVersionCheck() {
        let request = BGAppRefreshTaskRequest(identifier: taskIdentifier)
        request.earliestBeginDate = Date(timeIntervalSinceNow: 15 * 60) // 15 минут

        do {
            try BGTaskScheduler.shared.submit(request)
        } catch {
            print("Failed to schedule task: \(error)")
        }
    }

    private func handleVersionCheck(task: BGAppRefreshTask) {
        scheduleVersionCheck() // Планируем следующую задачу

        task.expirationHandler = {
            task.setTaskCompleted(success: false)
        }

        Task {
            do {
                let updates = try await fetchLastVersion()
                guard let latestUpdate = updates.first else {
                    task.setTaskCompleted(success: true)
                    return
                }
                if isNewerVersion(latestUpdate.version, currentVersion) {
                    try await downloadAndStoreBundle(latestUpdate)
                }
                if isReleaseDateReached(latestUpdate.releaseDate) {
                    try await extractBundle()
                    notifyWebView(latestUpdate)
                }
                task.setTaskCompleted(success: true)
            } catch {
                task.setTaskCompleted(success: false)
            }
        }
    }

    private func fetchLastVersion() async throws -> [OTAUpdate] {
        let url = URL(string: "https://mydomain.com/otalive/lastversion")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode([OTAUpdate].self, from: data)
    }

    private func isNewerVersion(_ newVersion: String, _ currentVersion: String) -> Bool {
        return newVersion > currentVersion
    }

    private func isReleaseDateReached(_ releaseDate: String) -> Bool {
        let formatter = ISO8601DateFormatter()
        guard let date = formatter.date(from: releaseDate) else { return false }
        return Date() >= date
    }

    private func downloadAndStoreBundle(_ update: OTAUpdate) async throws {
        let url = URL(string: "https://mydomain.com\(update.bundleUrl)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let zipFile = storageDir.appendingPathComponent("newversion.zip")
        try data.write(to: zipFile)
        if update.sha256 != data.sha256() { throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "SHA256 mismatch"]) }
    }

    private func extractBundle() async throws {
        let zipFile = storageDir.appendingPathComponent("newversion.zip")
        let extractDir = storageDir.appendingPathComponent("new_version")
        try FileManager.default.createDirectory(at: extractDir, withIntermediateDirectories: true)
        try FileManager.default.unzipItem(at: zipFile, to: extractDir)
    }

    private func notifyWebView(_ update: OTAUpdate) {
        notifyListeners("newVersionAvailable", data: [
            "version": update.version,
            "description": update.description
        ])
    }

    @objc func applyUpdate(_ call: CAPPluginCall) {
        Task {
            do {
                pendingVersionPath = storageDir.appendingPathComponent("new_version").path
                webView?.load(URLRequest(url: URL(fileURLWithPath: "\(pendingVersionPath!)/index.html")))
                try await validateCheckpoints()
                call.resolve()
            } catch {
                rollback()
                notifyListeners("updateFailed", data: ["error": error.localizedDescription])
                sendErrorLog(error.localizedDescription)
                call.reject(error.localizedDescription)
            }
        }
    }

    private func validateCheckpoints() async throws {
        let config = try loadUpdateConfig()
        calledCheckpoints.removeAll()

        // Регистрируем JS функцию
        webView?.evaluateJavaScript(
            "window.registerCheckpoint = function(name, executionTime) {" +
            "    window.dispatchEvent(new CustomEvent('checkpoint', { detail: { name: name, executionTime: executionTime } }));" +
            "};" +
            "document.addEventListener('checkpoint', function(e) {" +
            "    Capacitor.Plugins.OTA.checkpoint(e.detail);" +
            "});"
        )

        // Ждем таймаут
        try await Task.sleep(nanoseconds: UInt64(config.timeoutCheckMillis) * 1_000_000)

        // Проверяем, все ли чекпоинты вызваны
        if calledCheckpoints.count != config.requiredCheckpoints.count {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not all checkpoints were called"])
        }

        for (checkpoint, expectedTime) in config.requiredCheckpoints {
            if !calledCheckpoints.contains(checkpoint) {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Checkpoint \(checkpoint) was not called"])
            }
        }

        // Успех
        currentVersion = URL(fileURLWithPath: pendingVersionPath!).lastPathComponent
        pendingVersionPath = nil
        notifyListeners("updateSuccess", data: [:])
    }

    @objc func checkpoint(_ call: CAPPluginCall) {
        guard let data = call.getObject("detail"),
              let name = data["name"] as? String,
              let executionTime = data["executionTime"] as? Int else {
            call.reject("Invalid checkpoint data")
            return
        }

        do {
            let config = try loadUpdateConfig()
            if let expectedTime = config.requiredCheckpoints[name] {
                if executionTime == expectedTime {
                    calledCheckpoints.insert(name)
                    call.resolve()
                } else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Checkpoint \(name) execution time mismatch"])
                }
            } else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unknown checkpoint: \(name)"])
            }
        } catch {
            rollback()
            notifyListeners("updateFailed", data: ["error": error.localizedDescription])
            sendErrorLog(error.localizedDescription)
            call.reject(error.localizedDescription)
        }
    }

    private func loadUpdateConfig() throws -> UpdateConfig {
        let configFile = storageDir.appendingPathComponent("new_version/update.config.json")
        let data = try Data(contentsOf: configFile)
        return try JSONDecoder().decode(UpdateConfig.self, from: data)
    }

    private func rollback() {
        webView?.load(URLRequest(url: URL(fileURLWithPath: "\(storageDir.path)/current_version/index.html")))
        pendingVersionPath = nil
    }

    @objc func rollBackUpdate(_ call: CAPPluginCall) {
        rollback()
        notifyListeners("updateRolledBack", data: [:])
        call.resolve()
    }

    private func sendErrorLog(_ error: String) {
        Task {
            var request = URLRequest(url: URL(string: "https://mydomain.com/otalive/log")!)
            request.httpMethod = "POST"
            request.httpBody = try? JSONEncoder().encode(["error": error, "timestamp": ISO8601DateFormatter().string(from: Date())])
            try? await URLSession.shared.data(for: request)
        }
    }
}

extension Data {
    func sha256() -> String {
        // Реализовать вычисление SHA256
        return ""
    }
}

File: android/src/main/java/team/devapp/capacitor/otalive/OtaLiveUpdaterPlugin.java
package team.devapp.capacitor.otalive;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import androidx.work.*;
import com.getcapacitor.*;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.getcapacitor.plugin.WebView;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.HashSet;
import java.util.List;
import java.net.URL;
import java.util.Map;
import java.util.Set;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.Phaser;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicReference;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.net.MalformedURLException;
import okhttp3.*;

@CapacitorPlugin(name = "OtaLiveUpdater")
public class OtaLiveUpdaterPlugin extends Plugin {
    private static final String TAG = "OtaLiveUpdater";
    private static final String WORK_NAME = "ota_version_check";
    private static final String PREFS_NAME = "ota_prefs";
    private static final String KEY_CURRENT_VERSION = "current_version";
    private final OkHttpClient client = new OkHttpClient();
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private String currentVersion = "1.0.0";
    private String pendingVersionPath = null;
    private final Gson gson = new Gson();
    private final Set<String> calledCheckpoints = new HashSet<>();
    private SharedPreferences prefs;
    private SharedPreferences prefsWV;
    SharedPreferences.Editor editor;
    private static final Phaser semaphoreReady = new Phaser(1);
    static class OTAUpdate {
        String version;
        String date;
        String releaseDate;
        String bundleUrl;
        String description;
        String sha256;
    }

    static class UpdateConfig {
        Map<String, Integer> requiredCheckpoints;
        long timeoutCheckMillis;
    }

    @Override
    public void load() {
        Log.d(TAG, "OTA Plugin loaded");
        prefs = getContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        this.prefsWV = this.getContext().getSharedPreferences(WebView.WEBVIEW_PREFS_NAME, Activity.MODE_PRIVATE);
        this.editor= prefsWV.edit();

        currentVersion = prefs.getString(KEY_CURRENT_VERSION, "1.0.0");
        Log.d(TAG, "Loaded currentVersion: " + currentVersion);
        scheduleVersionCheck();
    }

    private void scheduleVersionCheck() {
        Log.d(TAG, "Scheduling WorkManager task");
        Constraints constraints = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build();

        VersionCheckWorker.bridge=this.getBridge();
        VersionCheckWorker.plugin=this;

        PeriodicWorkRequest workRequest = new PeriodicWorkRequest.Builder(VersionCheckWorker.class, 15, TimeUnit.MINUTES)
                .setConstraints(constraints)
                .setBackoffCriteria(BackoffPolicy.LINEAR, 10, TimeUnit.MINUTES)
                .addTag(WORK_NAME)
                .build();

        WorkManager.getInstance(getContext())
                .enqueueUniquePeriodicWork(WORK_NAME, ExistingPeriodicWorkPolicy.REPLACE, workRequest);
        Log.d(TAG, "WorkManager task enqueued with ID: " + workRequest.getId());

        // Проверяем статус задачи
        WorkManager.getInstance(getContext())
                .getWorkInfoByIdLiveData(workRequest.getId())
                .observeForever(workInfo -> {
                    if (workInfo != null) {
                        Log.d(TAG, "WorkInfo state: " + workInfo.getState());
                    }
                });
    }

    public static class VersionCheckWorker extends Worker {
        private static final String TAG = "OTAWorker";
        private final OkHttpClient client = new OkHttpClient();
        private final Gson gson = new Gson();
        private final File storageDir;
        private String currentVersion;
        private final SharedPreferences prefs;
        private static Bridge bridge;

        private static OtaLiveUpdaterPlugin plugin;
        public VersionCheckWorker(Context context, WorkerParameters params) {
            super(context, params);
            this.storageDir = context.getFilesDir();
            this.prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            this.currentVersion = prefs.getString(KEY_CURRENT_VERSION, "1.0.0");
            Log.d(TAG, "Worker initialized with currentVersion: " + currentVersion);
        }

        @Override
        public Result doWork() {
            Log.d(TAG, "Worker started");
            try {
                List<OTAUpdate> updates = fetchLastVersion();
                Log.d(TAG, "Fetched updates: " + updates.size());
                if (updates.isEmpty()) return Result.success();
                OTAUpdate latestUpdate = updates.get(0);
                if (isNewerVersion(latestUpdate.version, currentVersion)) {
                    Log.d(TAG, "New version found: " + latestUpdate.version);
                    downloadAndStoreBundle(latestUpdate);
                }
                if (isReleaseDateReached(latestUpdate.releaseDate)) {
                    Log.d(TAG, "Release date reached, extracting bundle");
                    extractBundle();
                    plugin.notifyWebView(latestUpdate);
                }
                return Result.success();
            } catch (Exception e) {
                Log.e(TAG, "Worker error: " + e.getMessage());
                return Result.retry();
            }
        }

        private List<OTAUpdate> fetchLastVersion() throws IOException {
            Log.d(TAG, "Fetching last version from server");
            Request request = new Request.Builder()
                    .url("http://192.168.1.231:5111/otalive/lastversion")
                    .build();
            Response response = client.newCall(request).execute();
            if (response.body() == null) {
                Log.e(TAG, "Empty response from server");
                throw new IOException("Empty response");
            }
            String json = response.body().string();
            Type listType = new TypeToken<List<OTAUpdate>>(){}.getType();
            return gson.fromJson(json, listType);
        }

        private boolean isNewerVersion(String newVersion, String currentVersion) {
            boolean isNewer = newVersion.compareTo(currentVersion) > 0;
            Log.d(TAG, "Comparing versions: new=" + newVersion + ", current=" + currentVersion + ", isNewer=" + isNewer);
            return isNewer;
        }

        private boolean isReleaseDateReached(String releaseDate) {
            try {
                boolean reached = System.currentTimeMillis() >= Long.parseLong(releaseDate);
                Log.d(TAG, "Release date check: " + releaseDate + ", reached=" + reached);
                return reached;
            } catch (NumberFormatException e) {
                Log.e(TAG, "Invalid release date format: " + releaseDate);
                return false;
            }
        }

        private void downloadAndStoreBundle(OTAUpdate update) throws IOException {
            Log.d(TAG, "Downloading bundle: " + update.bundleUrl);
            RequestBody reqbody = RequestBody.create(null, new byte[0]);
            Request request = new Request.Builder()
                    .url("http://192.168.1.231:5111/otalive/v/" + update.bundleUrl)
                    .post(reqbody)
                    .build();
            Response response = client.newCall(request).execute();
            if (response.body() == null) {
                Log.e(TAG, "Empty bundle response");
                throw new IOException("Empty bundle");
            }
            byte[] bytes = response.body().bytes();
            File zipFile = new File(storageDir, "newversion.zip");
            try (FileOutputStream fos = new FileOutputStream(zipFile)) {
                fos.write(bytes);
                fos.close();
            }

            String computedSha = sha256(bytes);
            if (!update.sha256.equals(computedSha)) {
                Log.e(TAG, "SHA256 mismatch: expected=" + update.sha256 + ", computed=" + computedSha);
                throw new IOException("SHA256 mismatch");
            }
            Log.d(TAG, "Bundle downloaded and stored");
        }

        private void extractBundle() throws IOException {
            Log.d(TAG, "Extracting bundle");
            File zipFile = new File(storageDir, "newversion.zip");
            File extractDir = new File(storageDir, "new_version");
            try (ZipFile zip = new ZipFile(zipFile)) {
                for (java.util.Enumeration<? extends ZipEntry> e = zip.entries(); e.hasMoreElements(); ) {
                    ZipEntry entry = e.nextElement();
                    File file = new File(extractDir, entry.getName());
                    if (entry.isDirectory()) {
                        file.mkdirs();
                    } else {
                        file.getParentFile().mkdirs();
                        try (InputStream is = zip.getInputStream(entry); FileOutputStream fos = new FileOutputStream(file)) {
                            byte[] buffer = new byte[1024];
                            int len;
                            while ((len = is.read(buffer)) > 0) {
                                fos.write(buffer, 0, len);
                            }
                        }
                    }
                }
            }
            Log.d(TAG, "Bundle extracted");
        }



        private String sha256(byte[] data) {
            try {
                MessageDigest digest = MessageDigest.getInstance("SHA-256");
                byte[] hash = digest.digest(data);
                StringBuilder hexString = new StringBuilder();
                for (byte b : hash) {
                    String hex = Integer.toHexString(0xff & b);
                    if (hex.length() == 1) hexString.append('0');
                    hexString.append(hex);
                }
                return hexString.toString();
            } catch (Exception e) {
                Log.e(TAG, "SHA256 error: " + e.getMessage());
                return "";
            }
        }
    }
    private void notifyWebView(OTAUpdate update) {
        Log.d(TAG, "Notifying WebView about new version: " + update.version);
        JSObject data = new JSObject();
        data.put("version", update.version);
        data.put("description", update.description);
        notifyListeners("newVersionAvailable",data,true);
        /*if (bridge != null) {

                    bridge.triggerJSEvent("newVersionAvailable", "OtaLiveUpdater", data.toString());
        }*/
    }

    @PluginMethod
    public void applyUpdate(PluginCall call) {

        executor.execute(() -> {
            try {
                pendingVersionPath = new File(getContext().getFilesDir(), "new_version").getAbsolutePath();

                editor.putString(WebView.CAP_SERVER_PATH, pendingVersionPath);
                editor.apply();
                editor.commit();
                _reload();
                // Переносим вызов WebView на главный поток
                /*getActivity().runOnUiThread(() -> {
                    Log.e(TAG,"UUUUUUUUUUURLLLL="+getBridge().getAppUrl());
                    Log.e(TAG,"UUUUUUUUUUURLLLL="+prefsWV.getString(WebView.CAP_SERVER_PATH,null));

                    //getBridge().setServerAssetPath("file://" + pendingVersionPath + "/index.html");
                    getBridge().getWebView().reload();
                    //getBridge().getWebView().loadUrl("file://" + pendingVersionPath + "/index.html");
                });*/
                validateCheckpoints();
                prefs.edit().putString(KEY_CURRENT_VERSION, currentVersion).apply();
                call.resolve();
            } catch (Exception e) {
                // Переносим rollback на главный поток
                getActivity().runOnUiThread(() -> {
                    rollback();
                });
                JSObject errorData = new JSObject();
                errorData.put("error", e.getMessage() != null ? e.getMessage() : "Unknown error");
                notifyListeners("updateFailed", errorData);
                sendErrorLog(e.getMessage() != null ? e.getMessage() : "Unknown error");
                call.reject(e.getMessage());
            }
        });
    }
    private void semaphoreUp() {
        Log.i(this.TAG, "semaphoreUp");
        semaphoreReady.register();
    }
    protected boolean _reload() {
        final String path = new File(getContext().getFilesDir(), "new_version").getAbsolutePath();///this.implementation.getCurrentBundlePath();
        this.semaphoreUp();
        Log.i(TAG, "Reloading: " + path);

        AtomicReference<URL> url = new AtomicReference<>();
        /*if (this.keepUrlPathAfterReload) {
            try {
                if (Looper.myLooper() != Looper.getMainLooper()) {
                    Semaphore mainThreadSemaphore = new Semaphore(0);
                    this.bridge.executeOnMainThread(() -> {
                        try {
                            url.set(new URL(this.bridge.getWebView().getUrl()));
                        } catch (Exception e) {
                            Log.e(CapacitorUpdater.TAG, "Error executing on main thread", e);
                        }
                        mainThreadSemaphore.release();
                    });
                    mainThreadSemaphore.acquire();
                } else {
                    try {
                        url.set(new URL(this.bridge.getWebView().getUrl()));
                    } catch (Exception e) {
                        Log.e(CapacitorUpdater.TAG, "Error executing on main thread", e);
                    }
                }
            } catch (InterruptedException e) {
                Log.e(CapacitorUpdater.TAG, "Error waiting for main thread or getting the current URL from webview", e);
            }
        }*/

        if (url.get() != null) {
            if (true){ //this.implementation.isUsingBuiltin()) {
                this.bridge.getLocalServer().hostAssets(path);
            } else {
                this.bridge.getLocalServer().hostFiles(path);
            }

            try {
                URL finalUrl = null;
                finalUrl = new URL(this.bridge.getAppUrl());
                finalUrl = new URL(finalUrl.getProtocol(), finalUrl.getHost(), finalUrl.getPort(), url.get().getPath());
                URL finalUrl1 = finalUrl;
                this.bridge.getWebView()
                        .post(() -> {
                            this.bridge.getWebView().loadUrl(finalUrl1.toString());
                            this.bridge.getWebView().clearHistory();
                        });
            } catch (MalformedURLException e) {
                Log.e(TAG, "Cannot get finalUrl from capacitor bridge", e);

                if (true/*this.implementation.isUsingBuiltin()*/) {
                    this.bridge.setServerAssetPath(path);
                } else {
                    this.bridge.setServerBasePath(path);
                }
            }
        } else {
            if (/*this.implementation.isUsingBuiltin()*/true) {
                this.bridge.setServerAssetPath(path);
            } else {
                this.bridge.setServerBasePath(path);
            }
        }

        //this.checkAppReady();
        this.notifyListeners("appReloaded", new JSObject());
        return true;
    }
    private void validateCheckpoints() throws Exception {
        Log.d(TAG, "Validating checkpoints");
        UpdateConfig config = loadUpdateConfig();
        calledCheckpoints.clear();
        CountDownLatch latch = new CountDownLatch(1);
/*
        getBridge().getWebView().evaluateJavascript(
                "window.registerCheckpoint = function(name, executionTime) {" +
                        "    window.dispatchEvent(new CustomEvent('checkpoint', { detail: { name: name, executionTime: executionTime } }));" +
                        "};",
                null
        );

        getBridge().getWebView().evaluateJavascript(
                "document.addEventListener('checkpoint', function(e) {" +
                        "    Capacitor.Plugins.OTA.checkpoint(e.detail.name, e.detail.executionTime);" +
                        "});",
                null
        );
*/
        executor.execute(() -> {
            try {
                Log.d(TAG, "Waiting for checkpoints, timeout: " + config.timeoutCheckMillis + "ms");
                Thread.sleep(config.timeoutCheckMillis);
                if (calledCheckpoints.size() < config.requiredCheckpoints.size()) {
                    throw new Exception("Not all checkpoints were called");
                }
            } catch (Exception e) {
                Log.e(TAG, "Checkpoint validation error: " + e.getMessage());
                rollback();
                JSObject errorData = new JSObject();
                errorData.put("error", e.getMessage());
                notifyListeners("updateFailed", errorData);
                sendErrorLog(e.getMessage());
            } finally {
                latch.countDown();
            }
        });

        latch.await(config.timeoutCheckMillis + 1000, TimeUnit.MILLISECONDS);

        if (calledCheckpoints.size() != config.requiredCheckpoints.size()) {
            Log.e(TAG, "Checkpoint validation failed: called " + calledCheckpoints.size() + ", required " + config.requiredCheckpoints.size());
            throw new Exception("Checkpoint validation failed: not all checkpoints called");
        }

        for (Map.Entry<String, Integer> entry : config.requiredCheckpoints.entrySet()) {
            if (!calledCheckpoints.contains(entry.getKey())) {
                Log.e(TAG, "Checkpoint not called: " + entry.getKey());
                throw new Exception("Checkpoint " + entry.getKey() + " was not called");
            }
        }

        currentVersion = new File(pendingVersionPath).getName();
        pendingVersionPath = null;
        Log.d(TAG, "Checkpoints validated, new currentVersion: " + currentVersion);
        notifyListeners("updateSuccess", new JSObject());
    }

    @PluginMethod
    public void checkpoint(PluginCall call) {
        String name = call.getString("name");
        Integer executionTime = call.getInt("executionTime");
        if (name == null || executionTime == null) {
            Log.e(TAG, "Invalid checkpoint data");
            call.reject("Invalid checkpoint data");
            return;
        }

        try {
            UpdateConfig config = loadUpdateConfig();
            if (config.requiredCheckpoints.containsKey(name)) {
                if (executionTime <= config.requiredCheckpoints.get(name)) {
                    calledCheckpoints.add(name);
                    Log.d(TAG, "Checkpoint called: " + name + ", executionTime: " + executionTime);
                    call.resolve();
                } else {
                    throw new Exception("Checkpoint " + name + " execution time mismatch");
                }
            } else {
                throw new Exception("Unknown checkpoint: " + name);
            }
        } catch (Exception e) {
            Log.e(TAG, "Checkpoint error: " + e.getMessage());
            rollback();
            JSObject errorData = new JSObject();
            errorData.put("error", e.getMessage());
            notifyListeners("updateFailed", errorData);
            sendErrorLog(e.getMessage());
            call.reject(e.getMessage());
        }
    }

    private UpdateConfig loadUpdateConfig() throws IOException {
        Log.d(TAG, "Loading update config");
        File configFile = new File(getContext().getFilesDir(), "new_version/update.config.json");
        try (java.io.FileReader reader = new java.io.FileReader(configFile)) {
            return gson.fromJson(reader, UpdateConfig.class);
        }
    }

    private void rollback() {
        editor.putString(WebView.CAP_SERVER_PATH, "public");
        editor.apply();
        editor.commit();

        // Выполняем загрузку URL на главном потоке
        getActivity().runOnUiThread(() -> {
            getBridge().getWebView().reload();
            pendingVersionPath = null;
        });
    }

    @PluginMethod
    public void rollBackUpdate(PluginCall call) {
        Log.d(TAG, "Rollback update requested");
        rollback();
        notifyListeners("updateRolledBack", new JSObject());
        call.resolve();
    }

    private void sendErrorLog(String error) {
        Log.d(TAG, "Sending error log: " + error);
        RequestBody body = RequestBody.create(
                MediaType.parse("application/json"),
                "{\"error\":\"" + error + "\",\"timestamp\":\"" + System.currentTimeMillis() + "\"}"
        );
        Request request = new Request.Builder()
                .url("http://192.168.1.231:5111/otalive/log")
                .post(body)
                .build();
        executor.execute(() -> {
            try {
                client.newCall(request).execute();
            } catch (IOException e) {
                Log.e(TAG, "Error log send failed: " + e.getMessage());
            }
        });
    }
}